rfcs/0001-extended-lookup-tables.md

```diff
@@ -0,0 +1,503 @@
+## Summary
+[summary]: #summary
+
+This RFC proposes an extension to our use of lookup tables using the PLOOKUP
+multiset inclusion argument, so that values within lookup tables can be chosen
+after the constraint system for a circuit has been fixed.
+
+## Motivation
+[motivation]: #motivation
+
+This extension should provide us with
+* array-style lookups (`arr[i]`, where `arr` and `i` are formed of values in
+  the proof witness).
+* the ability to load 'bytecode' using a proof-independent commitment.
+* the ability to load tables of arbitrary data for use in the circuit, allowing
+  e.g. the full data of a HTTPS transcript to be entered into a proof without
+  using some in-circuit commitment.
+* the ability to expose tables of arbitrary in-circuit data, for re-use by
+  other proofs.
+
+These goals support 5 major use-cases:
+* allow the verifier circuit to support 'custom' or user-specified gates
+  - using array indexing, the use of constants and commitments/evaluations can
+    be determined by the verifier index, instead of being fixed by the
+    permutation argument as currently.
+* allow random-access lookups in arrays as a base-primitive for user programs
+* allow circuits to use low-cost branching and other familiar programming
+  primitives
+  - this depends on the development of a bytecode interpreter on top of the
+    lookup-table primitives.
+* allow circuits to load and execute some 'user provided' bytecode
+  - e.g. to run a small bytecode program provided in a transaction at
+    block-production time.
+* allow zkOracles to load a full HTTPS transcript without using an expensive
+  in-circuit commitment.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+In order to support the desired goals, we first define 3 types of table:
+* **fixed tables** are tables declared as part of the constraint system, and
+  are the same for every proof of that circuit.
+* **runtime tables** are tables whose contents are determined at proving time,
+  to be used for array accesses to data from the witness.
+* **side-loaded tables** are tables that are committed to in a
+  proof-independent way, so that they may be reused across proofs and/or signed
+  without knowledge of the specific proof, but may be different for each
+  execution of the circuit.
+
+These distinct types of tables are a slight fiction: often it will be desirable
+to have some fixed part of runtime or side-loaded tables, e.g. to ensure that
+indexing is reliable. For example, a table representing an array of values
+`[x, y, z]` in the proof might be laid out as
```

Comment:
IIUC we also need to have different gates to support these: side-loaded table lookup gate, runtime table lookup gate, and as many fixed table lookup gate as we want to have.

Perhaps built-in table is a better name than fixed table?

Reply:
This RFC deliberately doesn't mention the actual gates, but we should be able to use the same gate for lookups in any of the tables, since they'll all be combined into 1 table.

For me, 'built-in' seems to imply something stronger, like it being an inherent part of kimchi itself. 'Baked-in' sits slightly better with me, but I think I still prefer 'fixed table'.